### PR TITLE
Cyclic replication

### LINKED ISSUE #471: Completely support cyclic replication in TiCDC
## Description

**Is your feature request related to a problem? Please describe:**

In many usage scenarios, The replication topology may exist loopback, as follows:

```
                   --> TiCDC --> 
Cluster-A                           Cluster-B
                   <-- TiCDC <--
```

```
Cluster-C <- TiCDC <- Cluster-A <- TiCDC <- Cluster-B <- TiCDC <- Cluster-C
```

**TODO list**

Following is the task lists of the whole design

- [x] #472 Mark table design and mark table operation in MySQL sink
- [x] #473 Support mark table subscription in processor
- [x] #474 Support cyclic traffic filter in MySQL sink
- [x] #475 Support column and data mismatch in cyclic replication, Divided into two sub problems, one is fixed, the other will be delayed (no good solution now), more details in https://github.com/pingcap/ticdc/issues/475#issuecomment-618218560

### Value description 

Strongly required by many scenarios.

### Value score 

* (TBD) / 5

### Workload estimation 

* need more tests
* 5 person-day in 7-28

### Time

GanttStart: 2020-06-01
GanttDue: 2020-07-30
GanttProgress: 90%

### PR DESCRIPTION
**Pull request by overvenus**
### What problem does this PR solve? 

Supports cyclic replication, resolves #471 .

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### Release note

- Support cyclic replication

### CODE REVIEW

**Comment by overvenus** [comment]
PTAL

File: pkg/filter/filter.go [hunk 1]
```diff
@@ -30,16 +38,45 @@ type DispatchRule struct {
 	Rule string `toml:"rule" json:"rule"`
 }
 
+// ReplicationConfig represents config used for cyclic replication
+type ReplicationConfig struct {
+	ReplicaID       uint64   `toml:"enable" json:"enable"`
```
Comment (amyangfei) [inline pkg/filter/filter.go:43]:
```suggestion
	ReplicaID       uint64   `toml:"replica-id" json:"replica-id"`
```

File: cdc/model/changefeed.go [hunk 2]
```diff
@@ -92,5 +92,16 @@ func (info *ChangeFeedInfo) Marshal() (string, error) {
...
 	err := json.Unmarshal(data, &info)
-	return errors.Annotatef(err, "Unmarshal data: %v", data)
+	if err != nil {
+		return errors.Annotatef(err, "Unmarshal data: %v", data)
+	}
+	// TODO(neil) find a better way to let sink know cyclic is enabled.
+	if info.Config != nil && info.Config.Cyclic.IsEnabled() {
+		cyclicCfg, err := info.Config.Cyclic.Marshal()
+		if err != nil {
+			return errors.Annotatef(err, "Unmarshal data: %v", data)
```
Comment (amyangfei) [inline cdc/model/changefeed.go:102]:
```suggestion
			return errors.Annotatef(err, "Marshal data: %v", data)
```

File: cdc/processor.go [hunk 3]
```diff
@@ -484,8 +488,31 @@ func (p *processor) removeTable(tableID int64) {
...
 // handleTables handles table scheduler on this processor, add or remove table puller
-func (p *processor) handleTables(ctx context.Context, oldInfo, newInfo *model.TaskStatus, checkpointTs uint64) {
+func (p *processor) handleTables(
+	ctx context.Context, oldInfo, newInfo *model.TaskStatus, checkpointTs uint64,
+) error {
 	removedTables, addedTables := diffProcessTableInfos(oldInfo.TableInfos, newInfo.TableInfos)
+	if p.changefeed.Config != nil && p.changefeed.Config.Cyclic.IsEnabled() {
+		// Make sure all normal tables and mark tables are paired.
+		schemaSnapshot := p.schemaStorage.GetLastSnapshot()
+		for _, tables := range [][]*model.ProcessTableInfo{addedTables, removedTables} {
```
Comment (amyangfei) [inline cdc/processor.go:498]: should the check separated between `addTables` and `removedTables`?

File: pkg/filter/filter.go (hunk 1 above)
Comment (leoppro) [reply pkg/filter/filter.go:43]: i think we should add an `enable` field into `ReplicationConfig`

File: pkg/cyclic/mark.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+			} else {
+				multiRows = append(multiRows, rows)
+			}
+			output[name] = multiRows
+		}
+	}
+	for name, events := range output {
+		// Per table order may lose during map stage, we need to sort events.
+		sort.Slice(events, func(i, j int) bool { return events[i][0].CommitTs < events[j][0].CommitTs })
+		output[name] = events
```
Comment (leoppro) [inline pkg/cyclic/mark.go:175]: this line can be removed

File: pkg/cyclic/replication.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	// concatenated by "," like: mode1,mode2
+	newMode := strings.Replace(oldMode, toRemove+",", "", -1)
+	newMode = strings.Replace(newMode, ","+toRemove, "", -1)
+	newMode = strings.Replace(newMode, toRemove, "", -1)
+	return newMode
+}
+
+// Cyclic ...
```
Comment (leoppro) [inline pkg/cyclic/replication.go:58]:
```suggestion
// Cyclic ...
```
the comment of `Cyclic` is missing

File: pkg/cyclic/replication.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+// RelaxSQLMode returns relaxed SQL mode, "STRICT_TRANS_TABLES" is removed.
+func RelaxSQLMode(oldMode string) string {
+	toRemove := "STRICT_TRANS_TABLES"
+
+	if !strings.Contains(oldMode, toRemove) {
+		return oldMode
+	}
+
+	// concatenated by "," like: mode1,mode2
+	newMode := strings.Replace(oldMode, toRemove+",", "", -1)
```
Comment (leoppro) [inline pkg/cyclic/replication.go:52]: why not using `split`?
what happened if sql mode contains a space? like "STRICT_TRANS_TABLES ,SOME_THINE"

**Pull request merged**